File: ruby/lib/record.rb
# A record is a set of fields found in a record-based file
#
# ==== Example
#
#   # create new record
#   r = Record.new("RECORD1", "Description of record 1")
#
#   # add field objects
#   r.push(Field.new("FIELD1", "Description of field 1", FieldType.new("A", "string"), 10))
#   r.push(Field.new("FIELD2", "Description of field 2", FieldType.new("N", "integer"), 5))
#   r.push(Field.new("FIELD2", "Description of field 2", FieldType.new("N", "integer"), 5))
#   r.push(Field.new("FIELD3", "Description of field 3", FieldType.new("A", "string"), 10))
#
#   # access field by index
#   f1 = r[0]           #=> first field in the record
#
#   # access by field name
#   f2s = r["FIELD2"]   #=> array of Field objects whose name is "FIELD2"
#   f2s.size            #=> 3
#   
#   # test if field is in record
#   r.include?("FIELD1")   #=> true
#   r.include?("FOO")      #=> false
#
#   # get record data
#   r.size                 #=> 4
#   r.length               #=> 30
#
#   # set record value by setting all fields values
#   s = "AAAAAAAAAA0000100002BBBBBBBBBB"
#   r = s
#   r[0].value             #=> "AAAAAAAAAA"
#   r[1].value             #=> "00001"
#   r[2].value             #=> "00002"
#   r[3].value             #=> "BBBBBBBBBB"
#
#   # and build array of attributes
#   r.array_of(:@name)     #=> ["FIELD1","FIELD2","FIELD2","FIELD3"]
#   r.array_of(:@length)   #=> [10, 5, 5, 10]
#
#   # iterate on all fields
#    r.each {|field| puts field.name }
#
class Record < Element
    include Enumerable
    
   # attr_reader :flist, :fmap

    # create a record
    # :call-seq:
    # new(name, description) -> new_Record
    def initialize(name, description)
        # record name should not be empty
        raise ArgumentError, "record name is empty!" unless name != ""

        super(name, description, 0)

        # initialize array and hash which will store field objects
        @flist = Array.new
        @fmap = Hash.new {|h,k| h[k] = Array.new }
    end

    # Append a field into a record. At the same time, update field lower/upper bounds and other field
    # data only relevant when inside a record
    #
    # :call-seq:
    # push(Field) -> ary
    #
    # ==== Arguments
    #
    # * +field+ - a Field class object
    def <<(field)
        field.index = @flist.count
        field.offset = @length 

        # save field in a list, and in a hash because field might have the same name. Using the hash,
        # it's easier to retrieve
        @flist.push(field)
        @fmap[field.name].push(field)

        # at the same time, build alternate field name
        field.alternate_name = field.name + @fmap[field.name].size.to_s

        # as we add a field, record length is growing
        @length += field.length

        # and adjust field bounds
        field.lower_bound = field.offset
        field.upper_bound = field.offset + field.length
    end

    # Returns the number of fields in the record
    # :call-seq:
    # count -> fixnum
    def count
        @flist.count
    end

    # Field  access with either a string (field name) or an integer (field index). If a field name is passed,
    # as fields could have the same name, an array of Field objects is returned.
    #
    # :call-seq:
    # [](key) -> Field or ary of Field objects
    #
    # ==== Arguments
    #
    # * +key+ - return the field object matching field name (if key is a string), or its index in the record
    # if key is an integer
    #
    # === Raise
    #
    # +ArgumentError+ if:
    # * key (string) is not found
    # * key (integer) is out of bounds
    # * key is neither a string nor an integer
    def [](key)
        if key.is_a?(String)
            if !@fmap.include?(key) 
                raise ArgumentError, "key #{key} is not found in record #{name}!"
            end
            @fmap[key]
        elsif key.is_a?(Integer)
            if key < 0 or key >= self.count
                raise ArgumentError, "key #{key} is out of bounds in record #{name}!"
            end
            @flist[key]
        else
            raise ArgumentError, "key #{key} is not a string nor an integer for record #{name}!"
        end
    end

    # iterator: iterates through each field
    def each(&block)
        @flist.each(&block)
    end

    # set record value by setting all field values
    def value=(s)
        # adjust length to exactly fit record length
        if s.length < @length
            s = s.ljust(@length)
        elsif s.length > @length
            s = s[0...@length]
        end

        # set each field value because we already know field bounds
        self.each {|f| f.value = s[f.lower_bound...f.upper_bound] }

    end

    # return true if a field name is found in the record
    # :call-seq:
    # include?(fname) -> bool
    #
    # ==== Arguments
    #
    # +fname+ - field name
    def include?(fname)
        @fmap.include?(fname)
    end

    # build an array of attributes data for the argument
    # :call-seq:
    # array_of(attr) -> ary
    def array_of(attr)
        @flist.map {|x| x.instance_variable_get(attr) }
    end

    # return the record value, which is just the concatenation of all field raw values.
    # Returned string length is equal to record length
    #
    # :call-seq:
    # value -> str
    #
    def value
        self.array_of(:@raw_value).join
    end
end

File: python/rbf/record.py
"""
Defines a generic record made of fields and allow easy construction and manipulation of
underlying fields.

"""

from rbf.element import Element
from rbf.field import Field

class Record(Element):
    """
    Define a new record based on name and description.

    :param str name: name of the record
    :param str description: description of the record
    
    ::

        >>> from rbf.fieldtype import FieldType
        >>> from rbf.field import Field
        >>> from rbf.record import Record
        >>> rec = Record("RECORD1", "Description for record #1")
        >>> rec.name
        'RECORD1'
        >>> rec.description
        'Description for record #1'
        >>> ft = FieldType("A/N", "string")
        >>> rec.append(Field("FIELD1", "Description of field 1", ft, 10))
        >>> rec.append(Field("FIELD2", "Description of field 2", ft, 5))
        >>> rec.append(Field("FIELD2", "Description of field 2", ft, 5))
        >>> rec.append(Field("FIELD3", "Description of field 3", ft, 10))
        >>> rec.count()
        4
        >>> len(rec)
        30
        >>> rec[0]
        name=<FIELD1> description=<Description of field 1> length=<10> data_type_representation=<A/N> data_type_description=<string> type=<BaseType.string> value=<> raw_value=<> offset=<0> index=<0> bounds=<0:10>
        >>> rec["FIELD2"]
        [name=<FIELD2> description=<Description of field 2> length=<5> data_type_representation=<A/N> data_type_description=<string> type=<BaseType.string> value=<> raw_value=<> offset=<10> index=<1> bounds=<10:15>, name=<FIELD2> description=<Description of field 2> length=<5> data_type_representation=<A/N> data_type_description=<string> type=<BaseType.string> value=<> raw_value=<> offset=<15> index=<2> bounds=<15:20>]
        >>> len(rec["FIELD2"])
        2
        >>> rec.value = "AAAAAAAAAABBBBBCCCCCDDDDDDDDDD"
        >>> rec.FIELD1
        'AAAAAAAAAA'
        >>> len(rec.FIELD1)
        10
        >>> rec[2].value
        'CCCCC'
        >>> rec.array_of('name')
        ['FIELD1', 'FIELD2', 'FIELD2', 'FIELD3']
        >>> rec.array_of('value')
        ['AAAAAAAAAA', 'BBBBB', 'CCCCC', 'DDDDDDDDDD']
        >>> "FIELD1" in rec
        True
        >>> "FOO" in rec
        False

    """

    def __init__(self, name: str, description: str):
        # call parent class ctor
        super(self.__class__, self).__init__(name, description, 0)

        # will hold field objects as a list of fields and a dict whose
        # key is the field name and value a list of field objects
        self._field_list = []
        self._field_dict = {}

    @property
    def value(self):
        """

        the value of the record

        :getter: record value = string concatenation of all field values
        :setter: set the record value (= all field values) from its argument
        :type: str

        """
        return "".join(self.array_of('raw_value'))

    @value.setter
    def value(self, new_value):
        """
        set the value for this record by setting the value of all fields of this record

        """
        # test line length: if longer, cut it
        if len(new_value) > self.length:
            new_value = new_value[0:self.length]
        # or if longer strip it
        elif len(new_value) < self.length:
            new_value = new_value.ljust(self.length)

        # setting record value is setting value for all fields/records composing the record
        for field in self:
            # field value is a slice of initial string
            field.value = new_value[field.lower_bound:field.upper_bound]

    def __getitem__(self, key):
        """
        access to a field by key

        :param key: could be an int or slice (standard list item access) or a field which
                gives back list of Field objects matching name
        :type key: int or slice or str

        ::

            field = rec["FIELD1"]
            field = rec[1]
            field = rec[2:4]

        :warning: a record may contain fields having the same name. So
            accessing by name using [] returns an array of field objects (the same with a slice)

        :todo: check slice bounds
        :rtype: list or str

        """
        if type(key) == int or type(key) == slice:
            return self._field_list[key]
        elif type(key) == str:
            if key not in self._field_dict: 
                raise ValueError("key {0} not found in record {1}".format(key, self.name))
            return [f for f in self._field_list if f.name == key]
        else:
            raise TypeError("Object {0} is of type {1} and not in int or str".format(key, type(key)))

    def __contains__(self, key: str) -> bool:
        """
        test if given field name belongs to record

        :param str key: field name to check existence for

        ::

            if "FIELD1" in rec:
                print("FIELD1 is found!!")

        """
        return key in self._field_dict

    def __repr__(self) -> str:
        output = format(super(self.__class__, self).__repr__()) + "\n"
        for field in self:
            output += "\t{0}\n".format(field)
        return output

    def __iter__(self):
        """
        as the field list is held using a regular list, just
        returns its iterator

        ::

            for field in rec:
                print(field)


        """
        return iter(self._field_list)

    def __getattr__(self, name: str) -> str:
        """
        return the value of the field name

        :param str name: field name
        :raises ValueError: 
            * if **name** is not found in record
            * if **name** field occurence is > 1

        ::

            if rec.FIELD1 == "My value":
                print("FIELD1 value is ok!!")

        """
        if name not in self:
            raise ValueError("attribute name '{0}' not found in record {1}".format(name, self.name))

        # depending on whether it's a Field or a Record
        items = self[name]

        # more than 1 item? impossible to fetch something, too ambiguous
        if len(items) > 1:
            raise ValueError("Ambiguous attribute name {0}".format(name))

        return items[0].value

    def append(self, field: Field):
        """
        insert (at the end, so append) a field object into a record

        :param Field field: field object to append

        ::

            rec.append(Field("FIELD1", "Field description #1", FieldType("A","string"), 10))

        """

        # set field index at the same time
        field.index = len(self._field_list)

        # offset at this moment is merely the length of record (starts at 0)
        field.offset = self.length

        # add a field to record
        self._field_list.append(field)

        # as a field name can be appear more than once, we keep
        # track of all fields using this dict
        if field.name in self._field_dict:
            self._field_dict[field.name].append(field)
        else:
            self._field_dict[field.name] = [field]

        # record is becoming longer
        self.length += field.length

        # and adjust field bounds
        field.lower_bound = field.offset
        field.upper_bound = field.offset + field.length

    def count(self) -> int:
        """ return the number of fields in the record """
        return len(self._field_list)

    def array_of(self, attr_name: str) -> list:
        """ 
        build an array of field attributes data for the argument

        :param str attr_name: attribute name for building the list
        :raises ValueError: if **attr_name** is not a Field class attribute

        """
        if not hasattr(self[0], attr_name): 
            raise ValueError("Field class has no attribute named <{0}>!!".format(attr_name))

        return [getattr(f, attr_name) for f in self]

    def asdict(self, convert=True) -> dict:
        """
        build a dict where keys are field name and values are field value or typed_value
        for all fields composing the record

        :param bool convert: if True, conversion is done from string value to typed value
        """
        d = {}

        # loop through fields because insertion order is more useful
        for f in self:
            # do not store empty values
            if f.value == "": continue

            # try to convert if conversion is  needed
            if convert:
                converted_value = f.convert()
            else:
                converted_value = value

            # we already added a key: so need to change it to a list
            if f.name in d:
                if type(d[f.name]) == list:
                    d[f.name].append(converted_value)
                else:
                    old_value = d[f.name]
                    d[f.name] = [old_value, converted_value]
            else:
                d[f.name] = converted_value

        return d

    def delete(self, fname_list: list, stringent: bool = False):
        """
        delete all fields matching names passed in the **fname_list** list

        :param list fnames: list of field names to delete

        """
        for fname in fname_list:
            self._field_list = [f for f in self._field_list if f.name != fname]

            # don't care if field name is not found
            if stringent:
                del self._field_dict[fname]
            else:
                self._field_dict.pop(fname, None)
    
    def keep(self, fname_list: list):
        """ keep only the fields specified in the list

        :param list fnames: list of field names to keep. Others are deleted

        """
        self._field_dict = { k:v for k,v in self._field_dict.items() if k in fname_list }
        self._field_list = [f for f in self._field_list if f.name in fname_list]
    


